### PR TITLE
Add constraint connections

### PR DESCRIPTION
**Pull request by shans**
1. Add constraint connections to new recipes
1. Add shape construction and matching for new recipes
1. Add a strategy to deal with constraint connections
1. Make strategy-explorer more better

### CODE REVIEW

File: runtime/new-recipe.js [hunk 1]
```diff
@@ -146,6 +148,10 @@ class Particle extends Node {
     return this._connections[name];
   }
 
+  addConnectionNameIfMissing(name) {
```
Comment (dstockwell) [inline runtime/new-recipe.js]: ensureConnection?

File: runtime/new-recipe.js [hunk 2]
```diff
@@ -630,9 +680,19 @@ class Recipe {
     recipe._slots = this._slots.map(cloneTheThing);
     recipe._particles = this._particles.map(cloneTheThing);
 
+    recipe._connectionConstraints = this._connectionConstraints.map(cloneTheThing);
+
+    recipe._cloneMap = cloneMap;
```
Comment (dstockwell) [inline runtime/new-recipe.js]: todo, need to figure out how to not keep this on recipe

File: runtime/new-recipe.js (hunk 1 above)
Comment (shans) [inline runtime/new-recipe.js]: done

File: runtime/new-recipe.js (hunk 2 above)
Comment (shans) [inline runtime/new-recipe.js]: done

**Pull request merged**